File: LeetCode/C++/45. Jump Game II.cpp
#include <iostream>
#include <math.h>
#include <vector>
#include<algorithm>
using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        int minStep = 0;
		int nextTarget = 0;
		int nextNextTarget = 0;
		int len = nums.size();
		for (int step = 0; step < len;) {
			if (nextTarget >= len -1)
				return minStep;
			//计算下步过程中准备下下步
			while (step <= nextTarget) {
				nextNextTarget = max(nextNextTarget, step+nums[step]);
				step++;
			}
			nextTarget = nextNextTarget;
			minStep++;
		}
		return minStep;
    }
};                     

int main() {
	Solution so;
	vector<int> A;
	A.push_back(2);
	A.push_back(3);
	A.push_back(1);
	A.push_back(1);
	A.push_back(4);
	cout << so.jump(A) << endl;
	return 0;
}

File: LeetCode/Java/45. Jump Game II.java
package hard;

public class Jump_Game_II {

    public int jump(int[] nums) {
    	int minStep = 0;
    	int nextTarget = 0;
    	int nextNextTarget = 0;
    	for (int step = 0; step < nums.length;) {
    		if (nextTarget >= nums.length - 1) {
    			return minStep;
    		}
    		while (step <= nextTarget) {	//设定下个最远目标,途中又找下下个最远目标
    			nextNextTarget = Math.max(nextNextTarget, step + nums[step]);
    			step++;
    		}
    		nextTarget = nextNextTarget;	//把下下个目标变成下个目标
    		minStep++;
    	}
    	return minStep;
    }
	
	public static void main(String[] args) {
		Jump_Game_II main = new Jump_Game_II();
		int[] nums = {2,3,1,1,4};
		System.out.println(main.jump(nums));;
	}

}
